warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java

```diff
@@ -0,0 +1,144 @@
+package datawave.query.jexl.functions;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import datawave.data.type.Type;
+import datawave.query.attributes.Attribute;
+import datawave.query.attributes.AttributeFactory;
+import datawave.query.attributes.Document;
+import datawave.query.predicate.EventDataQueryFilter;
+import datawave.query.tld.TLD;
+import datawave.query.util.Tuple2;
+import datawave.query.util.TypeMetadata;
+import org.apache.accumulo.core.data.ArrayByteSequence;
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+public class EventFieldAggregator extends IdentityAggregator {
+    // speedy cache loading for types, duplicated from AttributeFactory with caching of types rather than classes
+    protected static LoadingCache<String,Type<?>> typeCache = CacheBuilder.newBuilder().maximumSize(128).expireAfterAccess(1, TimeUnit.HOURS)
+                    .build(new CacheLoader<String,Type<?>>() {
+                        @Override
+                        public Type<?> load(String clazz) throws Exception {
+                            Class<?> c = Class.forName(clazz);
+                            return (Type<?>) c.newInstance();
+                        }
+                    });
+    
+    private TypeMetadata typeMetadata;
+    private String defaultTypeClass;
+    
+    public EventFieldAggregator(String field, EventDataQueryFilter filter, int maxNextCount, TypeMetadata typeMetadata, String defaultTypeClass) {
+        super(Collections.singleton(field), filter, maxNextCount);
+        
+        this.typeMetadata = typeMetadata;
+        this.defaultTypeClass = defaultTypeClass;
+    }
+    
+    @Override
+    protected List<Tuple2<String,String>> parserFieldNameValue(Key topKey) {
+        String cq = topKey.getColumnQualifier().toString();
+        int nullIndex1 = cq.indexOf('\u0000');
+        String field = cq.substring(0, nullIndex1);
+        String value = cq.substring(nullIndex1 + 1);
+        
+        int dataTypeEnd = -1;
+        ByteSequence cf = topKey.getColumnFamilyData();
+        for (int i = 0; i < cf.getBackingArray().length; i++) {
```

Comment:
The valid bytes in cf.getBackingArray() only go through cf.length().  The backing array could be longer.

Reply:
good catch